### PR TITLE
storagenode/{blobstore,pieces}: remember the earliest timestamp of a trash piece

### PR DESCRIPTION
**Pull request by liori**
GC runs (roughly) once a week per satellite, moving deleted blobs to trash with a single timestamp. Then, for the whole week, we do not expect any new trashed blobs. So let's just remember what's the timestamp of the earliest piece to delete, and skip any trash runs before that timestamp.

Change-Id: Idbb86aac080455625ee526309a34f70a689e8864

Please describe the performance impact: hopefully positive.

## Code Review Checklist (to be filled out by reviewer)
 - [ ] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/main/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/main/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

**Comment by thepaul** [comment]
@littleskunk do you think this would address your concerns from #6609, in place of https://review.dev.storj.io/c/storj/storj/+/12413 ? It's hard for me to say, not running any large nodes myself.

**Comment by littleskunk** [comment]
If I read this code correctly it would reduce the cleanup run from 7 times in 7 days down to about 2 times. On the first run, it will find out what the lowest timestamp is and sleep until that time is reached. It doesn't persist the timestamp so any restart means one extra run.

At the same time, it is a cleaver and easy solution. Can we maybe combine it with the solution you worked on? Especially the part about sleeping until the date is reached is something we might want to merge.

**Comment by liori** [comment]
It's probably not a problem to persist `minTimestamp` as a next step. I didn't do that because I didn't think of it at the time.

**Review by thepaul — APPROVED** [review]
I think it'd be better to do this without persisting anything for now.

File: storagenode/pieces/trashchore.go [hunk 1]
```diff
@@ -80,9 +92,14 @@ func (chore *TrashChore) Run(ctx context.Context) (err error) {
 
 				chore.log.Info("emptying trash started", zap.Stringer("Satellite ID", satellite))
 				trashedBefore := time.Now().Add(-chore.trashExpiryInterval)
-				err := chore.store.EmptyTrash(ctx, satellite, trashedBefore)
+				minPieceTimestamp, err := chore.store.EmptyTrash(ctx, satellite, trashedBefore)
 				if err != nil {
 					chore.log.Error("emptying trash failed", zap.Error(err))
+				} else {
+					chore.log.Info("emptying trash finished", zap.Stringer("Satellite ID", satellite), zap.Time("Next viable run", minPieceTimestamp))
```
Comment (liori) [inline storagenode/pieces/trashchore.go:99]: Log message is incorrect.
```suggestion
					chore.log.Info("emptying trash finished", zap.Stringer("Satellite ID", satellite), zap.Time("Next viable run", minPieceTimestamp.Add(chore.trashExpiryInterval)))
```

**Review by thepaul — APPROVED** [review]

File: storagenode/blobstore/filestore/dir.go [hunk 2]
```diff
@@ -447,9 +447,10 @@ func (dir *Dir) TryRestoreTrashBlob(ctx context.Context, ref blobstore.BlobRef)
 // EmptyTrash walks the trash files for the given namespace and deletes any
 // file whose mtime is older than trashedBefore. The mtime is modified when
 // Trash is called.
-func (dir *Dir) EmptyTrash(ctx context.Context, namespace []byte, trashedBefore time.Time) (bytesEmptied int64, deletedKeys [][]byte, err error) {
+func (dir *Dir) EmptyTrash(ctx context.Context, namespace []byte, trashedBefore time.Time) (bytesEmptied int64, deletedKeys [][]byte, minPieceTimestamp time.Time, err error) {
```
Comment (thepaul) [inline storagenode/blobstore/filestore/dir.go:450]: Just as a minor thing, here at the blobstore layer we try to talk about things in terms of _blobs_ and not _pieces_. The blobstore doesn't know what a piece is. Can we change the name of this return value to `minBlobTimestamp` everywhere that it appears under `storagenode/blobstore/`?

**Review by profclems — APPROVED** [review]

**Pull request merged**